eZ/Publish/API/Repository/Iterator/BatchIterator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Iterator;
+
+use Iterator;
+
+final class BatchIterator implements Iterator
+{
+    public const DEFAULT_BATCH_SIZE = 25;
+
+    /** @var \eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter */
+    private $adapter;
+
+    /** @var Iterator|null */
+    private $innerIterator;
+
+    /** @var mixed|null */
+    private $current = null;
+
+    /** @var int */
+    private $position = 0;
+
+    /** @var int */
+    private $batchSize;
+
+    public function __construct(
+        BatchIteratorAdapter $adapter,
+        int $batchSize = self::DEFAULT_BATCH_SIZE
+    ) {
+        $this->adapter = $adapter;
+        $this->batchSize = $batchSize;
+    }
+
+    public function current()
+    {
+        return $this->current;
+    }
+
+    public function next(): void
```

Comment:
`Iterator::next()` is not always called first.

See https://www.php.net/manual/en/class.iterator.php for `foreach` loop:

```php
string(18) "myIterator::rewind"
string(17) "myIterator::valid"
string(19) "myIterator::current"
string(15) "myIterator::key"
int(0)
string(12) "firstelement"

string(16) "myIterator::next"
string(17) "myIterator::valid"
string(19) "myIterator::current"
string(15) "myIterator::key"
int(1)
string(13) "secondelement"

string(16) "myIterator::next"
string(17) "myIterator::valid"
string(19) "myIterator::current"
string(15) "myIterator::key"
int(2)
string(11) "lastelement"

string(16) "myIterator::next"
string(17) "myIterator::valid"
```

So you have to assume that if `Iterator::current()`, `Iterator::key()` or `Iterator::valid()` methods are called, then inner iterator should be implicitly initialized if it's not already.